npm/cli
# fix: warn message on engine mismatch

## Overview

Messages for mismatching engine fields are not printed to users currently, [contrary to the documented behavior](https://docs.npmjs.com/files/package.json#engines), this PR fixes it by making sure that warning messages from packages containing mismatch engines are actually printed at `WARN` level for users of the cli.

## :pencil2: Changes
- Fixed setting engine check warnings to each package
- Added proper catching of package warnings during validation phase
- Added tap test that validates print of engine warn msgs


## :link: References
- Bug report: https://npm.community/t/engines-and-engines-strict-ignored/4792
- Documented behavior: https://docs.npmjs.com/files/package.json#engines

## :camera_flash: Screenshots

**Before**

<img width="842" alt="Screen Shot 2019-09-26 at 11 23 53 AM" src="https://user-images.githubusercontent.com/220900/65701952-356f4180-e050-11e9-933f-54750b5c2a4e.png">

**After**

<img width="842" alt="Screen Shot 2019-09-26 at 11 24 11 AM" src="https://user-images.githubusercontent.com/220900/65701971-3acc8c00-e050-11e9-8799-535d1f917de5.png">


## :mag: Testing

_Manual testing_
1. Setup a package with some invalid `engines.node` value, eg:
```json
{
  "name": "engine-module",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "engines": {
    "node": ">=100"
  }
}

```
2. Try to install from it, eg: `npm install ./engine-module`
3. Should be able to observer **WARN** messages of type `notsup`

_Automated testing_

- unit test at `test/tap/validate-tree-check-warnings.js` checking that package warnings are correctly collected to top level tree
- integration test at `test/tap/check-engine-reqs.js` validating that warning messages are printed to users upon `npm install`

✅ This change has unit test coverage
✅ This change has integration test coverage

## :fire: Rollback

> If we observe something wrong with this change in production, how should we respond?

This can be reverted at any time
